README.md
## Sckitam plugin for VCV Rack

This page provides some information about the following modules:

* **[2DRotation](#2DRotation)**: Utility, 2D Rotation of 2 input signals
* **[2DAffine](#2DAffine)**: Utility, 2D Affine Transform of 2 intput signals
* **[FIFOQueue](#FIFOQueue)**: First-in First-out queue
* **[MarkovSeq](#MarkovSeq)**: Sequencer & Switch, 8 steps sequencers based on Markov chain
* **[PolygonalVCO](#PolygonalVCO)**: VCO based on the paper: C. Hohnerlein, M. Rest, and J. O. Smith III, “Continuousorder polygonal waveform synthesis,” in Proceedings of theInternational Computer Music Conference, Utrecht, Netherlands, 2016.
* **[WDelay](#WDelay)**: High precision delay line. Building block for Waveguide and Karplus-Strong synthesis as well as Flanger or Chorus effects. 


![](doc/SckitamVCV_v2.png)


## 2DRotation <a id="2DRotation"> </a>
![](doc/2DRotation.png)

The two input signals, X and Y, are considered as the horizontal and vertical displacements on a 2D plane. They form a curve. The module mix them following a rotation rule. The two output signals are defined by:

Out1 = cos(\theta) X - sin(\theta) Y

Out2 = sin(\theta) X + cos(\theta) Y

The **Angle** value in the range +/-pi is defined by the knob on the left. The angle can be CV modulated. The range of CV from +/-5V corresponds to +/-pi. 

The **Velocity** of the rotation can be specified by the knob on the right. The range of rotation values has two modes. In the low (L) velocity mode, the velocity can reach +/-90 rotations per second. In the high (H) velocity mode, the rotation is specified by a 1V/Oct rule. If the knob is at 0, the velocity corresponds to 261.63 rotations per second (C4). Each increment of +/-1 will mutiply/divide the number of rotations per second by a factor of 2. The velocity parameter can also be CV modulated.  

The two sliders below \Delta X (\Delta Y) respectively define a horizontal (vertical) translation before and after the rotation. This allows one to precisely position the curve in the 2D space if necessary. 

X & Y input and output are polyphonic. The actual number of channels is defined by the X input port.

## 2DAffine <a id="2DAffine"> </a>
![](doc/2DAffine.png)

Same as the [2DRotation](#2DRotation) without the velocity parameter but with two additional shearing parameters, Sx and Sy. The shearing is applied before the rotation:

Out1 = cos(\theta) (X + Sx Y) - sin(\theta) (Sy X + Y)

Out2 = sin(\theta) (X + Sx Y) + cos(\theta) (Sy X + Y)

The shearing creates horizontal and vertical deformations of the curve in the 2D plane. Note that the shearing can modify very significantly the signal dinamic. As a result, a (parabolic) saturator has been introduced at the output.  

X & Y input and output are polyphonic. The actual number of channels is defined by the X input port.

## FIFOQueue <a id="FIFOQueue"> </a>
![](doc/FIFOQueue.png)

Triple First-in First-out Queue. 

**TrigW**: An input trigger stores the content of any of the three input available at the **In** ports in the queue. The current number of items stored in the queue is shown on the right. The maximum number of items that can be stored is 128. When this value is reached any TrigW is ignored. 

**TrigR**: An input trigger extracts the first available item in each queue and makes it available at the **Out** ports. The number of item in the queue is decreased by one. If the queue is empty, nothing happens (in particular no **TrigOut** is generated).

**Loop**: A positive value will activate the looping. Any value extracted from the queue will be stored back as last item in the queue. 

**Clear**: A trigger will empty the three queues. 

**TrigOut**: Each time an item is extracted from the queue, a trigger is generated at this input. Note: If the queue is empty and one tries to read an item, nothing happens and no **TrigOut** is generated.

**O**: Output 10v (or 0v) if the queue is empty (or not). 

## MarkovSeq <a id="MarkovSeq"> </a>
![](doc/MarkovSeq.png)

**8 steps sequencer or 8 to 1 switch**: This module defines 8 states and each incoming clock generates a **new state** which is defined by probabilities associated to the **current state**. This is essentially a first order 
[Markov Chain](https://en.wikipedia.org/wiki/Markov_chain) and allows one to define a graph of events with associated transition probabilities. The transition probabilites from each current state, S1,...,S8, are defined by the central part of the module. 

![](doc/MarkovSeq_State_Transition.png)

Each slider specifies the relative probability to go from the current state to one of the next state. Note that the probabilities are relative in the sense that they are normalized by the sum of the 8 values associated to the possible transitions. As a result, if all sliders have the same value, all possible next states are equiprobable (independently of the actual value the sliders may have). As a special case, if all slider values are equal to 0, the context menu allows one to choose either to remain in the current state or to assume that transitions to any states are equiprobable. 

As the number of probabilities to specify is 64, their definition can be cumbersome. Presets can be used. However beside presets, "-" and "+" buttons have been introduced to support the probability definition task. Above each set of sliders, 4 buttons are provided. The example below shows the case of state S2:  
![](doc/MarkovSeq_Reset_probability.png)

The two buttons on the left of S2 modifies the probabilities of all states (Si) that may come to state 2. The "-" button sets these probabilities to 0. This means that S2 will not be reached any more. The "+" button sets the probabilities to 0.5. 

The two buttons on the right of S2 modifies the probabilities of S2 to go to all possible states (Si). The "-" button sets these probabilities to 0. The "+" button sets these probabilities to 0.5. 

The probabilities initialization is done such that the sequencer moves as a regular sequencer in the forward direction: 

![](doc/MarkovSeq_Initialization.png)

The states are visualized on the right side of the module: 

![](doc/MarkovSeq_States.png)

The current state is shown in red and the next state (the one that will be triggered by the next clock gate) in blue. The next state can also be forced either manually with the associated button or by a gate signal. 

The left part of the module defines the values that are used to generate the output value for each current state. 

![](doc/MarkovSeq_input.png)

All input ports and the output port are polyphonic (the actual number of channels is defined by the input port of state 0). For each current state, the output is the sum of the incoming signal at the corresponding input port and the value of the knob (the transitions between values defined by the knobs can be smoothed by a Slew knob).
If the input is polyphonic, the value of the knob value is added to all channels. The module can therefore be used as a sequencer, an 8 to 1 switch or a combination of both. The final output can also be scaled by a Gain parameter. 

The current state value (value between 0 and 7) is also available at one output port. Finally, for each transition,
a trigger is sent to the Trig output port of the current state: 
 
![](doc/MarkovSeq_States_output_triggers.png)

## PolygonalVCO <a id="PolygonalVCO"> </a>
![](doc/PolygonalVCO.png)

This is an oscillator based on the sampling of a 2D curve, here a polygon of order N, by a rotating phasor. It generates two output signals X and Y corresponding to the projection of the sampling on the horizontal and vertical axes. 

The VCO was proposed in C. Hohnerlein, M. Rest, and J. O. Smith III, “Continuous order polygonal waveform synthesis,” in Proceedings of theInternational Computer Music Conference, Utrecht, Netherlands, 2016; and further discussed in C. Hohnerlein, M. Rest and J. Parker, "Efficient anti-aliasing of a complex polygonal oscillator", 20th International Conference on Digital Audio Effects (DAFx17), Edinburgh, Scotland, 2017.

The oscillator has essentially two parameters: 

1. **N: Order of the polygon**. Its value ranges from 2.1 up to 20 and can be CV modulated. High values of N create polygons that converge towards a circle and the resulting waveforms are cosine (X) and sine (Y) waves. Lower N values generate signals with richer harmonic content. If one wants to restrict the possible set of harmonic variations, the N parameter can be quantized. The context menu provides several quantization step options: "None", "0.25", "0.33", "0.50" or "1.00"
2. **T: Teeth**: This parameter creates "teeth" shapes on the corners of the polygon. These shapes increase the harmonic content of the signals.    

The oscillator is polyphonic. The actual number of channels is defined by the pitch signal. The FM signal is assumed to have as many channels as the pitch signal. The N and T CV ports *may* also be polyphonic. If they are monophonic, their value is applied on all channels (otherwise individual values are used for each channel).
 
In order to minimize the CPU load, no visual representation of the curve is included in the module itself. However, as shown in the figure above, the use of an additional scope is highly recommended, at least for sound design.

## WDelay <a id="WDelay"> </a>
![](doc/WDelay.png)

This module is a simple building block for Digital Waveguide and Karplus-Strong synthesis. It can also be used to create Flanger or Chorus. It is a high precision delay where the output is computed with 3rd order Lagrange interpolation. Beside the main output, a secondary output called "pickup" is also available. It allows one to extract the signal at an arbitrary position of the delay buffer. 

#### Input 
1. "In": Input port of the delay.  
2. "FBack": Input port for the feedback signal. Note that, by default, the output signal is not routed back to the input of the delay. A connection has to be explicitely done. One of the motivation of this module is to allow experimenting with different filters (or possibly other type of processing) inserted in the feedback loop. In the context of waveguide synthesis, another delay may be inserted in the feedback loop. The FBack signal is scaled by a knob (values between 0 and 0.999) and may be mutipled by 1 or -1 before being fed back to the delay. 

#### Output
1. "Out": Main output of the delay.
2. "Pickup": Signal extracted at an arbirary position of the delay buffer. A knob specifies the position: 0 corresponds to the input of the delay buffer and 1 to the output of the delay buffer. The position can be CV modulated.   

#### Delay specification
The delay can be defined in two different ways that can be selected by the context menu "Delay specification....".

1. Pitch: The Delay knob defines the length of the delay so that a given pitch is produced in the context of a Karplus-Strong or waveguide synthesis approach. The input port is interpreted as 1V/Oct. 
2. Time: In this case, the delay is directly defined in ms (values between 0.1ms up to 1.0s). The modulation is interpreted as: Delay*(1+CVvalue/5).

Finally, the "DAdj" (delay adjustment) is provided for fine tuning of the delay time. It allows one to shorten the delay time by an arbitrary value corresponding to a time between 0 and 16 sample periods. This is particularily useful in the "Pitch" mode. As the feedback is not created by default, the simplest connection involves a cable connecting the output port to the feedback input port. In VCV, this cable implies one sample delay. At high frequencies, this extra sample delay may have a significant impact on the pitch. Therefore, the "DAdj" can be set to -1 to shorten the length of the delay so that the complete scheme "Delay + cable" has exactly the intended length. If filters or other processing blocks are inserted in the feedback loop, their delay should be compensated if a precise Pitch definition is necessary. 

In the following example: 

![](doc/Karplus-Strong-example.png)

[Patch available here](doc/Karplus_Strong.vcv)

the feedback loop involves the "Stabile" filter as well as a "DC" blocking filter (They are marked as "Filter 1" and "Filter 2"). The connection involves 3 cables. So the DAdjust should be at least of -3. In fact, a fine tuning with a tuner shows that the filters themselves add a small amount of extra delay. The final delay adjustment is set to -3.284. A VCO was added to check the tuning of the Karplus-Strong voice compared to a pure sine wave.  





CHANGELOG.md
#### 2.0.1 (2021-12-06)
- Added ByPass functonality for 2DRotation, 2DAffine and WDelay

#### 2.0.0 (2021-10-31)
- Update for VCV Rack v2
- **FIFOQueue**: Triple First-in First-out queue

#### 1.0.5 (2020-05-31)
- **WDelay**: Initial release of Waveguide delay

#### 1.0.4 (2020-02-27)
- **MarkovSeq**: GUI update (States from 1 to 8)
- Fix tags and module descriptions

#### 1.0.3 (2020-02-05)
- **Polygonal VCO**: Improved aliasing reduction for high teeth values and correction of bug related to FM input

#### 1.0.3 (2020-02-05)
- **Polygonal VCO**: Improved aliasing reduction for high teeth values and correction of bug related to FM input

#### 1.0.3 (2020-02-05)
- **Polygonal VCO**: Improved aliasing reduction for high teeth values and correction of bug related to FM input

#### 1.0.2 (2020-01-25)
- **All Modules**: Add support for polyphony
- **All Modules**: Add option in the context menu for light or dark panel
- **Polygonal VCO**: Add option in the context menu to quantize the N parameter

#### 1.0.1 (2020-01-14)
- **2DAffine**: Add parabolic saturation at the output as the shearing may increase a lot the signal dynamic

- **MarokSeq**: 
	1. Add output triggers for each state
	2. Add tools to set probabilities related to a given state to 0.0 or to 0.5
	3. Modify the initialization such that the sequencer moves as a regular sequencer (forward direction)
	4. Add a context menu to choose the behavior if all probabilities are 0

 
#### 1.0.0 (2020-01-08)
- Initial Release of 2DAffine, 2DRotation, MarkovSeq, PolygonalVCO


src/PolygonalVCO.cpp
#include "plugin.hpp"

struct PolygonalVCO : Module {
	enum ParamIds {
		PITCH_PARAM,
		FMAMOUNT_PARAM,
		NPOLY_PARAM,
		TEETH_PARAM,
		NPOLYAMOUNT_PARAM,
		TEETHAMOUNT_PARAM,
		NUM_PARAMS
	};
	enum InputIds {
		PITCH_INPUT,
		FM_INPUT,
		NPOLYCV_INPUT,
		TEETHCV_INPUT,
		NUM_INPUTS
	};
	enum OutputIds {
		X_OUTPUT,
		Y_OUTPUT,
		NUM_OUTPUTS
	};
	enum LightIds {
		NUM_LIGHTS
	};

	enum NPolyQuant{
		NONE,
		Q025,
		Q033,
		Q050,
		Q100,
		NUM_NPOLYQUANT
	};
	NPolyQuant	nPolyQuant;

	float 	pitch, PPhi[16]={}, Phi=0.f, modPhi[16]={}, modPhi1[16]={}, P=0.f, freq, NPoly, x, y, tmp1, tmp2;
	float	Teeth=0.f, fracDelay, PhiTr, fxprime, fyprime, fx, fy, Correction = 0.f, Gain = 1.f;
	float	xout0[16]={}, xout1[16]={}, xout2[16]={}, xout3[16]={}, xout[16]={};
	float	yout0[16]={}, yout1[16]={}, yout2[16]={}, yout3[16]={}, yout[16]={};
	float	h0, h1, h2, h3, d1, d2, d3, d4, d5;


	void onReset() override {
		nPolyQuant = NONE;
	}

	PolygonalVCO() {
		config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);
		configParam(PITCH_PARAM, -3.f, 3.f, 0.f, "Pitch (Octave)");
		configParam(FMAMOUNT_PARAM, 0.f, 1.f, 0.f, "FM amount");
		configParam(NPOLY_PARAM, 2.1f, 20.f, 4.f, "Number of polygon sides");
		configParam(TEETH_PARAM, 0.f, 1.f, 0.f, "Teeth value");
		configParam(NPOLYAMOUNT_PARAM, -1.f, 1.f, 0.f, "CV amount for N");
		configParam(TEETHAMOUNT_PARAM, -1.f, 1.f, 0.f, "CV amount for T");
		
		configInput(PITCH_INPUT, "Pitch");
		configInput(FM_INPUT, "FM (LTZ)");
		configInput(NPOLYCV_INPUT, "Number of polygon sides CV");
		configInput(TEETHCV_INPUT, "Teeth value CV");
		configOutput(X_OUTPUT, "VCO X");
		configOutput(Y_OUTPUT, "VCO Y");
		
		onReset();
	}

	void process(const ProcessArgs& args) override {
		
		// Use the pitch input to get number of channels
		int channels  = std::max(inputs[PITCH_INPUT].getChannels(), 1);
		int channelsN = std::max(inputs[NPOLYCV_INPUT].getChannels(), 1);
		int channelsT = std::max(inputs[TEETHCV_INPUT].getChannels(), 1);

		for (int c = 0; c < channels; c++) {
			// Compute the frequency from the pitch parameter and input
        		pitch  = params[PITCH_PARAM].getValue() + inputs[PITCH_INPUT].getVoltage(c);

			pitch  = clamp(pitch, -4.f, 4.f);
        		freq   = dsp::FREQ_C4 * std::pow(2.f, pitch);
			freq  += 700.f * params[FMAMOUNT_PARAM].getValue() * inputs[FM_INPUT].getVoltage(c); 
			freq   = clamp(freq,-20000.f,20000.f);

			// Compute NPoly value
			if (channelsN == 1){
				NPoly   = params[NPOLY_PARAM].getValue() 
					+ params[NPOLYAMOUNT_PARAM].getValue() * inputs[NPOLYCV_INPUT].getVoltage(); 
			}
			else {
				NPoly   = params[NPOLY_PARAM].getValue() 
					+ params[NPOLYAMOUNT_PARAM].getValue() * inputs[NPOLYCV_INPUT].getVoltage(c); 
			}

			switch(nPolyQuant) {
				case 0:
            				NPoly   = clamp(NPoly, 2.1f, 20.f);
            				break;
				case 1:
					NPoly   = round(4.f*NPoly)/4.f;
					NPoly   = clamp(NPoly, 2.25f, 20.f);
            				break;
				case 2:
					NPoly   = round(3.f*NPoly)/3.f;
					NPoly   = clamp(NPoly, 2.33333f, 20.f);
            				break;
				case 3:
            				NPoly   = round(2.f*NPoly)/2.f;
					NPoly   = clamp(NPoly, 2.5f, 20.f);
            				break;
				case 4:
					NPoly   = round(NPoly);
					NPoly   = clamp(NPoly, 3.f, 20.f);
            				break;
             			default: break;
			}

			// Compute the Teeth value
			if (channelsT == 1){
				Teeth = params[TEETH_PARAM].getValue() 
					+ 0.1f * params[TEETHAMOUNT_PARAM].getValue() * inputs[TEETHCV_INPUT].getVoltage(); 
			}
			else {
				Teeth = params[TEETH_PARAM].getValue() 
					+ 0.1f * params[TEETHAMOUNT_PARAM].getValue() * inputs[TEETHCV_INPUT].getVoltage(c); 
			}
			Teeth = clamp(Teeth,0.f, 1.f);

        		// Adapt the Teeth value to NPoly
			if (NPoly < 5.f){
				Teeth = Teeth * (-0.03f*NPoly*NPoly+0.4f*NPoly-0.66f);
			}
			else{
				Teeth = Teeth * (-0.0019f*NPoly*NPoly+0.07f*NPoly+0.2875f);
			}	
		
			// Accumulate the phase
        		PPhi[c] += freq * args.sampleTime;
        		if (PPhi[c] >= 0.5f)
           			PPhi[c] -= 1.f;
			else if (PPhi[c] < -0.5f)
	 	 		PPhi[c] += 1.f;
			Phi = 2.f * M_PI * PPhi[c];

			// Compute modPhi
			modPhi[c] += NPoly * freq * args.sampleTime;
		
			// Increment modPhi and Check if there is a discontinuity, if necessary compute data for PolyBlep and PolyBlam
			Correction = 0.f;
			if (modPhi[c] >= 1.0f){
				fracDelay = (modPhi[c] - 1.f)/(modPhi[c] - modPhi1[c]);
				tmp1 	  = 2.f * M_PI * freq * args.sampleTime;
				PhiTr 	  = Phi - fracDelay * tmp1;
				if (Teeth > 0.f){
					tmp2 = std::cos(M_PI/NPoly) * (1.f/std::cos(-1.f*M_PI/NPoly+Teeth)-1.f/std::cos(M_PI/NPoly+Teeth));
					fx   = std::cos(Phi) * tmp2;
					fy   = std::sin(Phi) * tmp2;
				}
				tmp2 = std::cos(M_PI/NPoly) * (std::tan(Teeth-(M_PI/NPoly))/std::cos(Teeth-(M_PI/NPoly)) 
						            -  std::tan(Teeth+(M_PI/NPoly))/std::cos(Teeth+(M_PI/NPoly)));
				//tmp2 = -2.f * std::tan(M_PI/NPoly); 
				fxprime    = tmp2 * std::cos(PhiTr) * tmp1;
				fyprime    = tmp2 * std::sin(PhiTr) * tmp1;
				Correction = 1.0;
				modPhi[c] = modPhi[c] - std::floor(modPhi[c]);
			}
			else if (modPhi[c] < 0.0f){
				fracDelay = (modPhi[c])/(modPhi[c] - modPhi1[c]);
				tmp1 	  = 2.f * M_PI * freq *  args.sampleTime;
				PhiTr 	  = Phi - fracDelay * tmp1;
				if (Teeth > 0.f){
					tmp2 = std::cos(M_PI/NPoly) * (1.f/std::cos(-M_PI/NPoly+Teeth)-1.f/std::cos(M_PI/NPoly+Teeth));
					fx   = - std::cos(Phi) * tmp2;
					fy   = - std::sin(Phi) * tmp2;
				}			
				tmp2 = -1.f * std::cos(M_PI/NPoly) * (std::tan(Teeth-(M_PI/NPoly))/std::cos(Teeth-(M_PI/NPoly)) 
						            -  std::tan(Teeth+(M_PI/NPoly))/std::cos(Teeth+(M_PI/NPoly)));
				//tmp2 = 2.f * std::tan(M_PI/NPoly); 
				fxprime    = tmp2 * std::cos(PhiTr) * tmp1;
				fyprime    = tmp2 * std::sin(PhiTr) * tmp1;
				Correction = 1.0;
				modPhi[c] = modPhi[c] - std::floor(modPhi[c]);

			}

			// Create Polygon
			P = std::cos(M_PI/NPoly)/std::cos((2*modPhi[c]-1)*M_PI/NPoly+Teeth);
			x = std::cos(Phi)*P;
			y = std::sin(Phi)*P;

			xout0[c] = 0.f; yout0[c] = 0.f;
			xout1[c] = x;   yout1[c] = y;
			if (Correction == 1.f){
				d1 = fracDelay; d2 = d1*d1; d3 = d2*d1; d4=d3*d1; d5=d4*d1;

				// PolyBlep correction (if Teeth is not zero)
				if (Teeth > 0.f){
					h3 =  0.03871f * d4 + 0.00617f * d3 + 0.00737f * d2 + 0.00029f * d1 ;
					h2 = -0.11656f * d4 + 0.14955f * d3 + 0.22663f * d2 + 0.18783f * d1 + 0.05254f;
					h1 =  0.11656f * d4 - 0.31670f * d3 + 0.02409f * d2 + 0.62351f * d1 - 0.5f;
					h0 = -0.03871f * d4 + 0.16102f * d3 - 0.25816f * d2 + 0.18839f * d1 - 0.05254f;

					xout0[c] = xout0[c] + fx*h0;
					xout1[c] = xout1[c] + fx*h1;
					xout2[c] = xout2[c] + fx*h2;
					xout3[c] = xout3[c] + fx*h3;
		    
					yout0[c] = yout0[c] + fy*h0;
					yout1[c] = yout1[c] + fy*h1;
					yout2[c] = yout2[c] + fy*h2;
					yout3[c] = yout3[c] + fy*h3;
				}

				// PolyBlam Correction
				h0 = -0.008333 * d5 + 0.041667 * d4 - 0.083333 * d3 + 0.083333 * d2 - 0.041667 * d1 + 0.008333;
				h1 =  0.025    * d5 - 0.083333 * d4 +                 0.333333 * d2 - 0.5      * d1 + 0.233333;
				h2 = -0.025    * d5 + 0.041667 * d4 + 0.083333 * d3 + 0.083333 * d2 + 0.041667 * d1 + 0.008333;
				h3 =  0.008333 * d5;

				xout0[c] = xout0[c] + fxprime*h0;
				xout1[c] = xout1[c] + fxprime*h1;
				xout2[c] = xout2[c] + fxprime*h2;
				xout3[c] = xout3[c] + fxprime*h3;
		
				yout0[c] = yout0[c] + fyprime*h0;
				yout1[c] = yout1[c] + fyprime*h1;
				yout2[c] = yout2[c] + fyprime*h2;
				yout3[c] = yout3[c] + fyprime*h3;
			}

			// Output
			Gain = 1.f - 0.5f * Teeth;
			xout[c] = clamp(2.5f * Gain * xout3[c], -10.f, 10.f);
			yout[c] = clamp(2.5f * Gain * yout3[c], -10.f, 10.f);

			// Keep values for the next iteration
			modPhi1[c] = modPhi[c];
			xout3[c] = xout2[c]; xout2[c] = xout1[c]; xout1[c] = xout0[c];
			yout3[c] = yout2[c]; yout2[c] = yout1[c]; yout1[c] = yout0[c];
		}

		// Set output values
		outputs[X_OUTPUT].setChannels(channels);
		outputs[Y_OUTPUT].setChannels(channels);
		outputs[X_OUTPUT].writeVoltages(xout);
		outputs[Y_OUTPUT].writeVoltages(yout);
	}	

	void setNPolyQuant(NPolyQuant nPolyQuant) {
		if (nPolyQuant == this->nPolyQuant)
			return;
		this->nPolyQuant = nPolyQuant;
	}

	json_t* dataToJson() override {
		json_t* rootJ = json_object();
		// N Quantization
		json_object_set_new(rootJ, "nPolyQuant", json_integer(nPolyQuant));
		return rootJ;
	}

	void dataFromJson(json_t* rootJ) override {
		json_t* nPolyQuantJ = json_object_get(rootJ, "nPolyQuant");
		// N Quantization
		if (nPolyQuantJ)
			nPolyQuant = (NPolyQuant) json_integer_value(nPolyQuantJ);
	}

};

struct NPolyQuantValueItem : MenuItem {
	PolygonalVCO* module;
	PolygonalVCO::NPolyQuant nPolyQuant;
	void onAction(const event::Action& e) override {
		module->setNPolyQuant(nPolyQuant);
	}
};


struct NPolyQuantItem : MenuItem {
	PolygonalVCO* module;
	Menu* createChildMenu() override {
		Menu* menu = new Menu;
		std::vector<std::string> nPolyQuantNames = {
			"None",
			"0.25",
			"0.33",
			"0.50",
			"1.00",
		};
		for (int i = 0; i < PolygonalVCO::NUM_NPOLYQUANT; i++) {
			PolygonalVCO::NPolyQuant nPolyQuant = (PolygonalVCO::NPolyQuant) i;
			NPolyQuantValueItem* item = new NPolyQuantValueItem;
			item->text = nPolyQuantNames[i];
			item->rightText = CHECKMARK(module->nPolyQuant == nPolyQuant);
			item->module = module;
			item->nPolyQuant = nPolyQuant;
			menu->addChild(item);
		}
		return menu;
	}
};


struct PolygonalVCOWidget : ModuleWidget {

	void appendContextMenu(Menu *menu) override {

		PolygonalVCO *module = dynamic_cast<PolygonalVCO*>(this->module);
		assert(module);

		// N Quantization
		menu->addChild(new MenuEntry);
		NPolyQuantItem* nPolyQuantItem = new NPolyQuantItem;
		nPolyQuantItem->text = "N: Quantization step";
		nPolyQuantItem->rightText = RIGHT_ARROW;
		nPolyQuantItem->module = module;
		menu->addChild(nPolyQuantItem);
	}	

	PolygonalVCOWidget(PolygonalVCO* module) {
		setModule(module);

		setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/PolygonalVCO.svg")));

		addChild(createWidget<ScrewSilver>(Vec(RACK_GRID_WIDTH, 0)));
		addChild(createWidget<ScrewSilver>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, 0)));
		addChild(createWidget<ScrewSilver>(Vec(RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));
		addChild(createWidget<ScrewSilver>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));

		addParam(createParamCentered<RoundBlackKnob>(mm2px(Vec(8.677, 27.007)), module, PolygonalVCO::PITCH_PARAM));
		addParam(createParamCentered<Trimpot>(mm2px(Vec(22.457,27.629)), module, PolygonalVCO::FMAMOUNT_PARAM));
		addParam(createParamCentered<RoundBlackKnob>(mm2px(Vec(8.677, 73.221)), module, PolygonalVCO::NPOLY_PARAM));
		addParam(createParamCentered<RoundBlackKnob>(mm2px(Vec(22.457, 73.221)), module, PolygonalVCO::TEETH_PARAM));
		addParam(createParamCentered<Trimpot>(mm2px(Vec(8.677, 85.505)), module, PolygonalVCO::NPOLYAMOUNT_PARAM));
		addParam(createParamCentered<Trimpot>(mm2px(Vec(22.457, 85.505)), module, PolygonalVCO::TEETHAMOUNT_PARAM));

		addInput(createInputCentered<PJ301MPort>(mm2px(Vec(8.677, 42.657)), module, PolygonalVCO::PITCH_INPUT));
		addInput(createInputCentered<PJ301MPort>(mm2px(Vec(22.457, 42.657)), module, PolygonalVCO::FM_INPUT));
		addInput(createInputCentered<PJ301MPort>(mm2px(Vec(8.677, 95.829)), module, PolygonalVCO::NPOLYCV_INPUT));
		addInput(createInputCentered<PJ301MPort>(mm2px(Vec(22.457, 95.829)), module, PolygonalVCO::TEETHCV_INPUT));

		addOutput(createOutputCentered<PJ301MPort>(mm2px(Vec(8.677, 115.516)), module, PolygonalVCO::X_OUTPUT));
		addOutput(createOutputCentered<PJ301MPort>(mm2px(Vec(22.457, 115.516)), module, PolygonalVCO::Y_OUTPUT));
	}
};


Model* modelPolygonalVCO = createModel<PolygonalVCO, PolygonalVCOWidget>("PolygonalVCO");
